nodejs/dist/uws.js

```diff
@@ -1,98 +1,131 @@
+        });
 
-    /* we currently do only support port as option */
-    this.nativeServer = new NativeServer(options.port);
-    var self = this;
+        this.nativeServer.onMessage((nativeSocket, message, binary) => {
+            const socket = this.nativeServer.getData(nativeSocket);
+            socket.emit('message', binary ? message : message.toString());
+        });
+    }
 
-    this.close = function () {
-        self.nativeServer.close();
+    /**
+     * Closes the server.
+     *
+     * @public
+     */
+    close() {
+        this.nativeServer.close();
+        this.nativeServer = undefined;
     }
-    this.broadcast = function (message, options) {
+
+    /**
+     * Broadcast a message to all sockets.
+     *
+     * @param {String|Buffer} message The message to broadcast
+     * @param {Object} options Broadcast options
+     * @public
+     */
+    broadcast(message, options) {
         /* only listen to binary option */
-        this.nativeServer.broadcast(message, options.binary);
-    };
-
-    /* original connection callback */
-    this.connectionCallback = function (nativeSocket) {
-        var socket = new Socket(nativeSocket, self);
-        self.nativeServer.setData(nativeSocket, socket);
-        self.emit('connection', socket);
-    };
-
-    /* this function needs to handle the upgrade directly! */
-    this.handleUpgrade = function (request, socket, upgradeHead, callback) {
+        this.nativeServer.broadcast(message, options && options.binary || false);
+    }
+
+    /**
+     * Handles a HTTP Upgrade request.
+     *
+     * @param {http.IncomingMessage} request HTTP request
+     * @param {net.Socket} Socket between server and client
+     * @param {Buffer} upgradeHead The first packet of the upgraded stream
+     * @param {Function} callback Callback function
+     * @public
+     */
+    handleUpgrade(request, socket, upgradeHead, callback) {
         /* register a special connection handler */
-        self.nativeServer.onConnection(function (nativeSocket) {
-            var socket = new Socket(nativeSocket, self);
-            self.nativeServer.setData(nativeSocket, socket);
+        this.nativeServer.onConnection((nativeSocket) => {
+            const sock = new Socket(nativeSocket, this);
+            this.nativeServer.setData(nativeSocket, sock);
 
             /* internal variables */
-            socket.upgradeReq = request;
-            socket.readyState = socket.OPEN;
-            callback(socket);
+            sock.upgradeReq = request;
+            callback(sock);
 
             /* todo: reset connection handler when the upgrade queue is empty */
             /* this will probably never be noticed as you don't mix upgrade handling */
         });
 
         /* upgrades will be handled immediately */
-        self.nativeServer.upgrade(socket._handle.fd, request.headers['sec-websocket-key']);
+        this.nativeServer.upgrade(socket._handle.fd, request.headers['sec-websocket-key']);
```

Comment:
Yes but I don't support that yet and it seems to be empty all the time. The Node.js documentation is not very clear about what it is but I guess it holds data received together with the HTTP upgrade request, which is always going to be empty as long as the client is not totally coco and sends both HTTP upgrade request and an WebSocket frame in the same go, without waiting for the response. That is at least how I understand it.